File: clojure/anagram/src/anagram.clj
(ns anagram
  (:require [clojure.string :as str]))

(defn count-letter
  [counter letter]
  (let [current (get counter letter)]
    (if current
      (assoc counter letter (inc current))
      (assoc counter letter 1))))

(defn letter-count
  [word]
  (reduce count-letter {} (str/split word #"")))

(defn is-anagram?
 [word candidate]
 (and
   (not= word (str/lower-case candidate))
   (= (letter-count word) (letter-count (str/lower-case candidate)))))

(defn anagrams-for
  [word candidates]
  (filter (partial is-anagram? (str/lower-case word)) candidates))

File: elixir/anagram/anagram.exs
defmodule Anagram do
  @doc """
  Returns all candidates that are anagrams of, but not equal to, 'base'.
  """
  @spec match(String.t, [String.t]) :: [String.t]
  def match(base, candidates) do
    reference = String.downcase(base)

    candidates
    |> Enum.filter(&(String.downcase(&1) != reference))
    |> Enum.filter(&(is_anagram?(&1, count_letters(reference))))
  end

  defp is_anagram?(word, base_count) do
    String.downcase(word)
    |> count_letters
    |> Kernel.==(base_count)
  end

  defp count_letters(word) do
    word
    |> String.split("", trim: true)
    |> Enum.reduce(%{}, fn(letter, counter) -> Map.update(counter, letter, 1, &(&1 + 1)) end)
  end
end
